File: wear/src/main/java/com/windkracht8/wearmusicplayer/CommsWifi.java
/*
 * Copyright 2024-2025 Bart Vullings <dev@windkracht8.com>
 * This file is part of WearMusicPlayer
 * WearMusicPlayer is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * WearMusicPlayer is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.windkracht8.wearmusicplayer;

import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.util.Log;

import androidx.annotation.NonNull;

import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.Socket;

class CommsWifi{
    static boolean isReceiving = false;
    static void receiveFile(Main main, String path, long length, String ip, int port){
        Log.d(Main.LOG_TAG, "CommsWifi path: " + path);
        Log.d(Main.LOG_TAG, "CommsWifi length: " + length);
        Log.d(Main.LOG_TAG, "CommsWifi ip: " + ip);
        Log.d(Main.LOG_TAG, "CommsWifi port: " + port);

        ConnectivityManager connectivityManager = (ConnectivityManager)main.getSystemService(Main.CONNECTIVITY_SERVICE);
        try{
            connectivityManager.requestNetwork(
                    new NetworkRequest.Builder()
                            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
                            .build(),
                    new ConnectivityManager.NetworkCallback(){
                        public void onAvailable(@NonNull Network network){
                            super.onAvailable(network);
                            Log.d(Main.LOG_TAG, "CommsWifi.NetworkCallback.onAvailable");
                            connectivityManager.bindProcessToNetwork(network);
                            main.commsConnectionInfo(R.string.connection_info_Wifi);
                            readFileFromStream(main, connectivityManager, path, length, ip, port);
                        }
                        public void onUnavailable(){
                            super.onUnavailable();
                            Log.d(Main.LOG_TAG, "CommsWifi.NetworkCallback.onUnavailable");
                            main.toast(R.string.fail_wifi_fast);
                            readFileFromStream(main, connectivityManager, path, length, ip, port);
                        }
                    },
                    2000
            );
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "CommsWifi " + e.getMessage());
            main.toast(R.string.fail_wifi_fast);
        }
    }
    private static void readFileFromStream(Main main, ConnectivityManager connectivityManager, String path, long length, String ip, int port){
        Log.d(Main.LOG_TAG, "CommsWifi.readFileFromStream");
        isReceiving = true;
        long bytesDone = 0;
        try(Socket socket = new Socket(ip, port)){
            InputStream inputStream = socket.getInputStream();
            try(FileOutputStream fileOutputStream = new FileOutputStream(Library.musicDir + path)){
                int counter = 0;
                while(counter < 100){
                    if(inputStream.available() == 0){
                        counter++;
                        sleep100();
                        continue;
                    }
                    counter = 0;
                    byte[] buffer = new byte[2048];
                    int numBytes = inputStream.read(buffer);
                    if(numBytes < 0){
                        Log.e(Main.LOG_TAG, "CommsWifi.receiveFile read error");
                        main.toast(R.string.fail_read_wifi);
                        main.commsFileFailed(path, R.string.fail_read_wifi);
                        connectivityManager.bindProcessToNetwork(null);
                        isReceiving = false;
                        return;
                    }
                    fileOutputStream.write(buffer, 0, numBytes);
                    bytesDone += numBytes;

                    long progress = (bytesDone * 100) / length;
                    main.commsProgress((int)progress);
                    if(bytesDone >= length){
                        main.commsFileDone(path);
                        connectivityManager.bindProcessToNetwork(null);
                        isReceiving = false;
                        return;
                    }
                }
                main.toast(R.string.fail_read_wifi);
                main.commsFileFailed(path, R.string.fail_read_wifi);
            }catch(Exception e){
                Log.e(Main.LOG_TAG, "CommsWifi.receiveFile FileOutputStream exception: " + e.getMessage());
                main.toast(R.string.fail_write_file);
                main.commsFileFailed(path, R.string.fail_write_file);
            }
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "CommsWifi.receiveFile Socket exception: " + e);
            Log.e(Main.LOG_TAG, "CommsWifi.receiveFile Socket exception: " + e.getMessage());
            main.toast(R.string.fail_wifi);
            main.commsFileFailed(path, R.string.fail_wifi);
        }
        //if we get here it failed
        connectivityManager.bindProcessToNetwork(null);
        isReceiving = false;
    }
    private static void sleep100(){
        try{
            Thread.sleep(100);
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "CommsWifi.sleep100 exception: " + e.getMessage());
        }
    }
}

File: wear/src/main/java/CommsWifi.kt
/*
 * Copyright 2024-2025 Bart Vullings <dev@windkracht8.com>
 * This file is part of WearMusicPlayer
 * WearMusicPlayer is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * WearMusicPlayer is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.windkracht8.wearmusicplayer

import android.app.Activity
import android.content.Context.CONNECTIVITY_SERVICE
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import java.io.FileOutputStream
import java.net.Socket

object CommsWifi {
	var connectivityManager: ConnectivityManager? = null
	var path: String = ""
	enum class Status { DONE, PREPARING, RECEIVING, ERROR }
	val status = MutableSharedFlow<Status>()
	enum class ConnectionType { REQUESTING, FAST, SLOW }
	var connectionType by mutableStateOf(null as ConnectionType?)
	var progress by mutableFloatStateOf(0F)
	var error = 0

	fun init(activity: Activity) {
		connectivityManager = activity.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager?
	}

	fun receiveFile(
		path: String,
		length: Long,
		ip: String,
		port: Int
	) {
		logD("CommsWifi path: $path length: $length ip: $ip port: $port")
		if(connectivityManager == null){
			error = R.string.fail_wifi
			CoroutineScope(Dispatchers.Default).launch { status.emit(Status.ERROR) }
			return
		}
		val connectivityManager = connectivityManager ?: return
		this.path = path
		CoroutineScope(Dispatchers.Default).launch { status.emit(Status.PREPARING) }
		connectionType = ConnectionType.REQUESTING
		progress = 0F

		try {
			connectivityManager.requestNetwork(
				NetworkRequest.Builder()
					.addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
					.build(),
				object : ConnectivityManager.NetworkCallback() {
					override fun onAvailable(network: Network) {
						super.onAvailable(network)
						logD("CommsWifi.NetworkCallback.onAvailable")
						connectivityManager.bindProcessToNetwork(network)
						CoroutineScope(Dispatchers.Default).launch {
							connectionType = ConnectionType.FAST
							readFileFromStream(connectivityManager, path, length, ip, port)
						}
					}
					override fun onUnavailable() {
						super.onUnavailable()
						logD("CommsWifi.NetworkCallback.onUnavailable")
						CoroutineScope(Dispatchers.Default).launch {
							connectionType = ConnectionType.SLOW
							readFileFromStream(connectivityManager, path, length, ip, port)
						}
					}
				},
				5000
			)
		} catch (e: Exception) {
			logE("CommsWifi " + e.message)
			CoroutineScope(Dispatchers.Default).launch {
				connectionType = ConnectionType.SLOW
				readFileFromStream(connectivityManager, path, length, ip, port)
			}
		}
	}

	suspend fun readFileFromStream(
		connectivityManager: ConnectivityManager,
		path: String,
		length: Long,
		ip: String,
		port: Int
	) {
		logD("CommsWifi.readFileFromStream")
		CoroutineScope(Dispatchers.Default).launch { status.emit(Status.RECEIVING) }
		var bytesDone: Long = 0
		try {
			Socket(ip, port).use { socket ->
				socket.getInputStream().use { inputStream ->
					FileOutputStream(Library.musicDir + path).use { fileOutputStream ->
						var lastReadTime = System.currentTimeMillis()
						while (System.currentTimeMillis() - lastReadTime < 3000) {
							if (inputStream.available() == 0) {
								delay(100)
								continue
							}
							val buffer = ByteArray(2048)
							val numBytes = inputStream.read(buffer)
							if (numBytes < 0) {
								logE("CommsWifi.receiveFile read error")
								error = R.string.fail_read_wifi
								CoroutineScope(Dispatchers.Default).launch { status.emit(Status.ERROR) }
								connectivityManager.bindProcessToNetwork(null)
								return
							}
							fileOutputStream.write(buffer, 0, numBytes)
							bytesDone += numBytes.toLong()
							progress = bytesDone.toFloat() / length.toFloat()
							//logD("CommsWifi.read bytesDone: $bytesDone length: $length progress: $progress")
							if (bytesDone >= length) {
								CoroutineScope(Dispatchers.Default).launch { status.emit(Status.DONE) }
								connectivityManager.bindProcessToNetwork(null)
								return
							}
							lastReadTime = System.currentTimeMillis()
						}
						error = R.string.fail_read_wifi
						CoroutineScope(Dispatchers.Default).launch { status.emit(Status.ERROR) }
					}
				}
			}
		} catch (e: Exception) {
			logE("CommsWifi.receiveFile exception: $e")
			logE("CommsWifi.receiveFile exception: " + e.message)
			error = R.string.fail_wifi
			CoroutineScope(Dispatchers.Default).launch { status.emit(Status.ERROR) }
		}
		//if we get here it failed
		connectivityManager.bindProcessToNetwork(null)
		error = R.string.fail_read_wifi
		CoroutineScope(Dispatchers.Default).launch { status.emit(Status.ERROR) }
	}
}
